2.13a1: doInBackgroundAnswerCard - RuntimeException

2.13a1 bugs out with "Collection not opened" for two of my four active decks (plus their parent):

Previous version was whatever had been on the Playstore alpha channel (I finally joined).

```
25574 I AnkiDroid: Obtaining card
21479 I AnkiDroid: AnkiActivity::onStart - Reviewer
25574 E AnkiDroid: CollectionTask/ doInBackgroundAnswerCard - RuntimeException on answering card
25574 E AnkiDroid: com.ichi2.libanki.WrongId:  No card with id 0
25574 E AnkiDroid:     at com.ichi2.libanki.Card.load(Card.java:24)
25574 E AnkiDroid:     at com.ichi2.libanki.Card.<init>(Card.java:8)
25574 E AnkiDroid:     at com.ichi2.libanki.Collection.getCard(Collection.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getLrnCard(SchedV2.java:8)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getLrnCard(SchedV2.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getCard(SchedV2.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2.getCard(SchedV2.java:4)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackgroundAnswerCard(CollectionTask.java:10)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:51)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
25574 E AnkiDroid:     at android.os.AsyncTask$3.call(AsyncTask.java:378)
25574 E AnkiDroid:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
25574 E AnkiDroid:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
25574 E AnkiDroid:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
25574 E AnkiDroid:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
25574 E AnkiDroid:     at java.lang.Thread.run(Thread.java:919)
25574 E AnkiDroid:
25574 E AnkiDroid: com.ichi2.libanki.WrongId:  No card with id 0
25574 E AnkiDroid:     at com.ichi2.libanki.Card.load(Card.java:24)
25574 E AnkiDroid:     at com.ichi2.libanki.Card.<init>(Card.java:8)
25574 E AnkiDroid:     at com.ichi2.libanki.Collection.getCard(Collection.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getLrnCard(SchedV2.java:8)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getLrnCard(SchedV2.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2._getCard(SchedV2.java:1)
25574 E AnkiDroid:     at com.ichi2.libanki.sched.SchedV2.getCard(SchedV2.java:4)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackgroundAnswerCard(CollectionTask.java:10)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:51)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
25574 E AnkiDroid:     at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
25574 E AnkiDroid:     at android.os.AsyncTask$3.call(AsyncTask.java:378)
25574 E AnkiDroid:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
25574 E AnkiDroid:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
25574 E AnkiDroid:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
25574 E AnkiDroid:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
25574 E AnkiDroid:     at java.lang.Thread.run(Thread.java:919)
```
and further
```
21479 I AnkiDroid: Handling Activity Result: 901. Result: 203
21479 I AnkiDroid: Creating notification channel with id/name: General Notifications/AnkiDroid
21479 I AnkiDroid: Creating notification channel with id/name: Synchronization/Synchronization
21479 I AnkiDroid: Creating notification channel with id/name: Global Reminders/Cards due
21479 I AnkiDroid: Creating notification channel with id/name: Deck Reminders/Reminders
21479 I AnkiDroid: Displaying Database Error
21479 W System.err: java.io.IOException: Cannot run program "sqlite3": error=2, No such file or directory
21479 W System.err:    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1050)
21479 W System.err:    at java.lang.Runtime.exec(Runtime.java:698)
21479 W System.err:    at java.lang.Runtime.exec(Runtime.java:528)
21479 W System.err:    at java.lang.Runtime.exec(Runtime.java:425)
21479 W System.err:    at com.ichi2.anki.dialogs.DatabaseErrorDialog.onCreateDialog(DatabaseErrorDialog.java:8)
21479 W System.err:    at com.ichi2.anki.dialogs.DatabaseErrorDialog.onCreateDialog(DatabaseErrorDialog.java:1)
21479 W System.err:    at androidx.fragment.app.DialogFragment.onGetLayoutInflater(DialogFragment.java:3)
21479 W System.err:    at androidx.fragment.app.Fragment.performGetLayoutInflater(Fragment.java:1)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:70)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManagerImpl.java:9)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:166)
21479 W System.err:    at androidx.fragment.app.BackStackRecord.executeOps(BackStackRecord.java:25)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.executeOps(FragmentManagerImpl.java:6)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:14)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManagerImpl.java:10)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:4)
21479 W System.err:    at androidx.fragment.app.FragmentManagerImpl.executePendingTransactions(FragmentManagerImpl.java:1)
21479 W System.err:    at com.ichi2.anki.AnkiActivity.showDialogFragment(AnkiActivity.java:6)
21479 W System.err:    at com.ichi2.anki.AnkiActivity.showAsyncDialogFragment(AnkiActivity.java:2)
21479 W System.err:    at com.ichi2.anki.AnkiActivity.showAsyncDialogFragment(AnkiActivity.java:1)
21479 W System.err:    at com.ichi2.anki.DeckPicker.showDatabaseErrorDialog(DeckPicker.java:2)
21479 W System.err:    at com.ichi2.anki.DeckPicker.handleDbError(DeckPicker.java:2)
21479 W System.err:    at com.ichi2.anki.DeckPicker.onActivityResult(DeckPicker.java:3)
21479 W System.err:    at android.app.Activity.dispatchActivityResult(Activity.java:8145)
21479 W System.err:    at android.app.ActivityThread.deliverResults(ActivityThread.java:4966)
21479 W System.err:    at android.app.ActivityThread.handleSendResult(ActivityThread.java:5014)
21479 W System.err:    at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
21479 W System.err:    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
21479 W System.err:    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
21479 W System.err:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2131)
21479 W System.err:    at android.os.Handler.dispatchMessage(Handler.java:107)
21479 W System.err:    at android.os.Looper.loop(Looper.java:214)
21479 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:7707)
21479 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
21479 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
21479 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
21479 W System.err: Caused by: java.io.IOException: error=2, No such file or directory
21479 W System.err:    at java.lang.UNIXProcess.forkAndExec(Native Method)
21479 W System.err:    at java.lang.UNIXProcess.<init>(UNIXProcess.java:133)
21479 W System.err:    at java.lang.ProcessImpl.start(ProcessImpl.java:141)
21479 W System.err:    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
21479 W System.err:    ... 35 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java
@@ -45,12 +45,10 @@ import com.ichi2.utils.JSONObject;
 import java.lang.ref.WeakReference;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.Comparator;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
-import java.util.ListIterator;
 import java.util.Locale;
 import java.util.Random;
 
@@ -378,15 +376,10 @@ public class Sched extends SchedV2 {
                            "SELECT due, id FROM cards WHERE did IN " + _deckLimit() + " AND queue = " + Consts.QUEUE_TYPE_LRN + " AND due < ? AND id != ? LIMIT ?",
                            new Object[]{mDayCutoff, currentCardId(), mReportLimit});
             while (cur.moveToNext()) {
-                mLrnQueue.add(new long[] { cur.getLong(0), cur.getLong(1) });
+                mLrnQueue.add(new LrnCard(cur.getLong(0), cur.getLong(1) ));
             }
             // as it arrives sorted by did first, we need to sort it
-            Collections.sort(mLrnQueue, new Comparator<long[]>() {
-                @Override
-                public int compare(long[] lhs, long[] rhs) {
-                    return Long.valueOf(lhs[0]).compareTo(rhs[0]);
-                }
-            });
+            Collections.sort(mLrnQueue);
             return !mLrnQueue.isEmpty();
         } finally {
             if (cur != null && !cur.isClosed()) {
@@ -409,8 +402,8 @@ public class Sched extends SchedV2 {
             if (collapse) {
                 cutoff += mCol.getConf().getInt("collapseTime");
             }
-            if (mLrnQueue.getFirst()[0] < cutoff) {
-                long id = mLrnQueue.remove()[1];
+            if (mLrnQueue.getFirst().getDue() < cutoff) {
+                long id = mLrnQueue.remove().getId();
                 Card card = mCol.getCard(id);
                 // mLrnCount -= card.getLeft() / 1000; See decrementCount()
                 return card;
@@ -481,7 +474,7 @@ public class Sched extends SchedV2 {
                 // it twice in a row
                 card.setQueue(Consts.QUEUE_TYPE_LRN);
                 if (!mLrnQueue.isEmpty() && mRevCount == 0 && mNewCount == 0) {
-                    long smallestDue = mLrnQueue.getFirst()[0];
+                    long smallestDue = mLrnQueue.getFirst().getDue();
                     card.setDue(Math.max(card.getDue(), smallestDue + 1));
                 }
                 _sortIntoLrn(card.getDue(), card.getId());
@@ -1664,16 +1657,16 @@ public class Sched extends SchedV2 {
      */
     @Override
     protected void _sortIntoLrn(long due, long id) {
-        Iterator i = mLrnQueue.listIterator();
+        Iterator<LrnCard> i = mLrnQueue.listIterator();
         int idx = 0;
         while (i.hasNext()) {
-            if (((long[]) i.next())[0] > due) {
+            if (i.next().getDue() > due) {
                 break;
             } else {
                 idx++;
             }
         }
-        mLrnQueue.add(idx, new long[] { due, id });
+        mLrnQueue.add(idx, new LrnCard(due, id));
     }
 
 

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
@@ -49,7 +49,6 @@ import java.lang.ref.WeakReference;
 import java.util.ArrayList;
 import java.util.Calendar;
 import java.util.Collections;
-import java.util.Comparator;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Iterator;
@@ -99,7 +98,26 @@ public class SchedV2 extends AbstractSched {
 
     // Queues
     protected final LinkedList<Long> mNewQueue = new LinkedList<>();
-    protected final LinkedList<long[]> mLrnQueue = new LinkedList<>();
+    protected class LrnCard implements Comparable<LrnCard> {
+        private final long mCid;
+        private final long mDue;
+        public LrnCard(long due, long cid) {
+            mCid = cid;
+            mDue = due;
+        }
+        public long getDue () {
+            return mDue;
+        }
+        public long getId() {
+            return mCid;
+        }
+
+        @Override
+        public int compareTo(LrnCard o) {
+            return Long.compare(mDue, o.mDue);
+        }
+    }
+    protected final LinkedList<LrnCard> mLrnQueue = new LinkedList<>();
     protected final LinkedList<Long> mLrnDayQueue = new LinkedList<>();
     protected final LinkedList<Long> mRevQueue = new LinkedList<>();
 
@@ -891,15 +909,10 @@ public class SchedV2 extends AbstractSched {
                             "SELECT due, id FROM cards WHERE did IN " + _deckLimit() + " AND queue IN (" + Consts.QUEUE_TYPE_LRN + ", " + Consts.QUEUE_TYPE_PREVIEW + ") AND due < ?"
                             + " AND id != ? LIMIT ?", new Object[] { cutoff, currentCardId(), mReportLimit});
             while (cur.moveToNext()) {
-                mLrnQueue.add(new long[] { cur.getLong(0), cur.getLong(1) });
+                mLrnQueue.add(new LrnCard(cur.getLong(0), cur.getLong(1)));
             }
             // as it arrives sorted by did first, we need to sort it
-            Collections.sort(mLrnQueue, new Comparator<long[]>() {
-                @Override
-                public int compare(long[] lhs, long[] rhs) {
-                    return Long.compare(lhs[0], rhs[0]);
-                }
-            });
+            Collections.sort(mLrnQueue);
             return !mLrnQueue.isEmpty();
         } finally {
             if (cur != null && !cur.isClosed()) {
@@ -921,8 +934,8 @@ public class SchedV2 extends AbstractSched {
             if (collapse) {
                 cutoff += mCol.getConf().getInt("collapseTime");
             }
-            if (mLrnQueue.getFirst()[0] < cutoff) {
-                long id = mLrnQueue.remove()[1];
+            if (mLrnQueue.getFirst().getDue() < cutoff) {
+                long id = mLrnQueue.remove().getId();
                 Card card = mCol.getCard(id);
                 // mLrnCount -= 1; see decrementCounts()
                 return card;
@@ -1089,7 +1102,7 @@ public class SchedV2 extends AbstractSched {
                 // sure we don't put it at the head of the queue and end up showing
                 // it twice in a row
                 if (!mLrnQueue.isEmpty() && mRevCount == 0 && mNewCount == 0) {
-                    long smallestDue = mLrnQueue.getFirst()[0];
+                    long smallestDue = mLrnQueue.getFirst().getDue();
                     card.setDue(Math.max(card.getDue(), smallestDue + 1));
                 }
                 _sortIntoLrn(card.getDue(), card.getId());
@@ -2807,16 +2820,16 @@ public class SchedV2 extends AbstractSched {
      * Sorts a card into the lrn queue LIBANKI: not in libanki
      */
     protected void _sortIntoLrn(long due, long id) {
-        Iterator i = mLrnQueue.listIterator();
+        Iterator<LrnCard> i = mLrnQueue.listIterator();
         int idx = 0;
         while (i.hasNext()) {
-            if (((long[]) i.next())[0] > due) {
+            if (i.next().getDue() > due) {
                 break;
             } else {
                 idx++;
             }
         }
-        mLrnQueue.add(idx, new long[] { due, id });
+        mLrnQueue.add(idx, new LrnCard(due, id));
     }
 
 

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
@@ -891,7 +891,7 @@ public class SchedV2 extends AbstractSched {
                             "SELECT due, id FROM cards WHERE did IN " + _deckLimit() + " AND queue IN (" + Consts.QUEUE_TYPE_LRN + ", " + Consts.QUEUE_TYPE_PREVIEW + ") AND due < ?"
                             + " AND id != ? LIMIT ?", new Object[] { cutoff, currentCardId(), mReportLimit});
             while (cur.moveToNext()) {
-                mLrnQueue.add(new long[] { cur.getLong(0), currentCardId(), cur.getLong(1) });
+                mLrnQueue.add(new long[] { cur.getLong(0), cur.getLong(1) });
             }
             // as it arrives sorted by did first, we need to sort it
             Collections.sort(mLrnQueue, new Comparator<long[]>() {

